File: py/idkollen_client/models/bankid_no.py
from __future__ import annotations

from typing import Annotated, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from idkollen_client.models.common import ApiErrorCode


class _Base(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class BankIdNoAuthRequest(_Base):
    redirect_url: Optional[str] = None
    request_ssn: Optional[bool] = None
    request_phone: Optional[bool] = None
    request_email: Optional[bool] = None
    request_address: Optional[bool] = None
    ref_id: Optional[str] = None
    app_callback_uri: Optional[str] = None


class BankIdNoBackchannelAuthRequest(_Base):
    ssn: str
    callback_url: Optional[str] = None
    ref_id: Optional[str] = None


class BankIdNoSignRequest(_Base):
    redirect_url: Optional[str] = None
    text: Optional[str] = None
    documents: Optional[list[str]] = None
    request_ssn: Optional[bool] = None
    request_phone: Optional[bool] = None
    request_email: Optional[bool] = None
    request_address: Optional[bool] = None
    ref_id: Optional[str] = None


class BankIdNoPending(_Base):
    status: Literal["PENDING"]
    id: str
    ref_id: Optional[str] = None
    url: Optional[str] = None
    binding_message: Optional[str] = None


class BankIdNoSignResult(_Base):
    end_user: str
    merchant: str
    hash: str


class BankIdNoSignedDocument(_Base):
    id: str
    hash: str


class BankIdNoCompleted(_Base):
    status: Literal["COMPLETED"]
    id: str
    ref_id: Optional[str] = None
    ssn: str
    name: str
    given_name: str
    surname: str
    phone: Optional[str] = None
    email: Optional[str] = None
    address: Optional[str] = None
    birth_date: Optional[str] = None
    pid: Optional[str] = None
    bank_id: Optional[str] = None
    sign_result: Optional[BankIdNoSignResult] = None
    signed_documents: Optional[list[BankIdNoSignedDocument]] = None


class BankIdNoFailed(_Base):
    status: Literal["FAILED"]
    id: str
    ref_id: Optional[str] = None
    error: ApiErrorCode


BankIdNoStatus = Annotated[
    BankIdNoPending | BankIdNoCompleted | BankIdNoFailed,
    Field(discriminator="status"),
]

File: csharp/src/Idkollen.Client/Models/BankIdNo.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Idkollen.Client.Models;

public sealed record BankIdNoAuthRequest
{
    public string? RedirectUrl { get; init; }
    public bool? RequestSsn { get; init; }
    public bool? RequestPhone { get; init; }
    public bool? RequestEmail { get; init; }
    public bool? RequestAddress { get; init; }
    public string? RefId { get; init; }
    public string? AppCallbackUri { get; init; }
}

public sealed record BankIdNoBackchannelAuthRequest(string Ssn)
{
    public string? CallbackUrl { get; init; }
    public string? RefId { get; init; }
}

public sealed record BankIdNoSignRequest
{
    public string? RedirectUrl { get; init; }
    public string? Text { get; init; }
    public List<string>? Documents { get; init; }
    public bool? RequestSsn { get; init; }
    public bool? RequestPhone { get; init; }
    public bool? RequestEmail { get; init; }
    public bool? RequestAddress { get; init; }
    public string? RefId { get; init; }
}

public sealed record BankIdNoSignResult(string EndUser, string Merchant, string Hash);
public sealed record BankIdNoSignedDocument(string Id, string Hash);

public interface IBankIdNoStatus { }

public sealed record BankIdNoPending(
    string Id,
    string? RefId,
    string? Url,
    string? BindingMessage
) : IBankIdNoStatus;

public sealed record BankIdNoCompleted(
    string Id,
    string? RefId,
    string Ssn,
    string Name,
    string GivenName,
    string Surname,
    string? Phone,
    string? Email,
    string? Address,
    string? BirthDate,
    string? Pid,
    string? BankId,
    BankIdNoSignResult? SignResult,
    List<BankIdNoSignedDocument>? SignedDocuments
) : IBankIdNoStatus;

public sealed record BankIdNoFailed(
    string Id,
    string? RefId,
    string Error
) : IBankIdNoStatus;

internal sealed class BankIdNoStatusConverter : JsonConverter<IBankIdNoStatus>
{
    public override IBankIdNoStatus Read(ref Utf8JsonReader reader, Type _, JsonSerializerOptions options)
    {
        using var doc = JsonDocument.ParseValue(ref reader);
        var status = doc.RootElement.GetProperty("status").GetString();
        var raw = doc.RootElement.GetRawText();
        return status switch
        {
            "PENDING" => JsonSerializer.Deserialize<BankIdNoPending>(raw, options)!,
            "COMPLETED" => JsonSerializer.Deserialize<BankIdNoCompleted>(raw, options)!,
            "FAILED" => JsonSerializer.Deserialize<BankIdNoFailed>(raw, options)!,
            _ => throw new JsonException($"Unknown bankid-no status: {status}"),
        };
    }

    public override void Write(Utf8JsonWriter w, IBankIdNoStatus v, JsonSerializerOptions o)
        => throw new NotSupportedException();
}
